### PR TITLE
[Ready for Review] Support more convenient foreach variable/value search-ability.

### PR DESCRIPTION
**Pull request by darinyu**
In this change, we

1. add values of foreach items to flowspec for convenient access to values in foreach steps. 
2. Update foreach stack to include value.
3. Add formatted foreach stack (path from root to current level) to metadata, guarded by a flag.

Currently,  we try to get "value" of the for each in the following order:
1. If the item is list of primitive types like `["a", "b", "c"]`, we will use the literal value directly
2. Otherwise, we use replib to get a serialized value of the iterable, limited to 30 charactors.

The end result will be of the form `[value]`.  We will then

1. Use the list of value (generated in a split/parent node) in the `task_finish` function inside a custom decorator.  
2. Use the list of value when constructing foreach stack, which will eventually be used to construct the for-loop path in metadata.

An example for the updated search UI is like the following.
Metaflow-UI (2)

### CODE REVIEW

File: metaflow/flowspec.py [hunk 1]
```diff
@@ -499,6 +501,33 @@ def _validate_ubf_step(self, step_name):
             )
             raise InvalidNextException(msg)
 
+    def _get_foreach_item_value(self, item):
```
Comment (romain-intel) [inline metaflow/flowspec.py:504]: nit: Add typing information here as well since you add it in the docs (I know we are not consistent but little by little :) )

File: metaflow/task.py [hunk 2]
```diff
@@ -188,10 +197,10 @@ def lineage():
                     if join_type == "foreach":
                         top = i["_foreach_stack"][-1]
                         bottom = i["_foreach_stack"][:-1]
-                        # the topmost indices in the stack are all
-                        # different naturally, so ignore them in the
+                        # the topmost indices and values in the stack are
+                        # all different naturally, so ignore them in the
                         # assertion
-                        yield bottom + [top._replace(index=0)]
+                        yield bottom + [top._replace(index=0, value=0)]
```
Comment (romain-intel) [inline metaflow/task.py:203]: shouldn't it be value=None?

File: metaflow/task.py [hunk 3]
```diff
@@ -463,6 +471,37 @@ def run_step(
             # 3. initialize foreach state
             self._init_foreach(step_name, join_type, inputs, split_index)
 
+            # Add foreach stack to metadata of the task
+
+            foreach_stack = (
```
Comment (romain-intel) [inline metaflow/task.py:476]: I've been thinking about this and trying to find a way to do this without actually storing a potentially large (and worse, unbounded) string in the metadata. I think we can probably do the following. In the metadata, we can store the split index instead. The `_foreach_stack` is actually available above as well as the `_foreach_values` so we would then be able, in the UI, to reconstruct the string that is being constructed below. This would be a tad be more complicated in the UI (but not that much and it would have the added benefit of being able to use the cache so fewer requests (ie: for each foreach nest, we would only need to load 2 artifacts to have all the paths for all the tasks in the foreach).

Thoughts?

File: metaflow/task.py [hunk 4]
```diff
@@ -188,10 +197,10 @@ def lineage():
                     if join_type == "foreach":
                         top = i["_foreach_stack"][-1]
                         bottom = i["_foreach_stack"][:-1]
-                        # the topmost indices in the stack are all
-                        # different naturally, so ignore them in the
+                        # the topmost indices and values in the stack are
```
Comment (romain-intel) [inline metaflow/task.py:200]: nit: I know this is not in your modification but what assert?

File: metaflow/runtime.py [hunk 5]
```diff
@@ -1027,7 +1027,13 @@ def results(self):
     @property
     def finished_id(self):
         # note: id is not available before the task has finished
-        return (self.step, tuple(self.results["_foreach_stack"]))
+        # index already identifies the task within the foreach,
```
Comment (romain-intel) [inline metaflow/runtime.py:1030]: why is this needed? Isn't the presence of the value backward compatible with the old code?

File: metaflow/flowspec.py (hunk 1 above)
Comment (darinyu) [reply metaflow/flowspec.py:504]: added

File: metaflow/runtime.py (hunk 5 above)
Comment (darinyu) [reply metaflow/runtime.py:1030]: for a node, we may want to quickly get the finish_id of the sibling node and check if they all finished (can check usage of `finished_id` within `_queue_task_join`). It is non-trivial to get value of the sibling, so we may as well not compare value at all (afterall, the index of the foreach already identifies the node).

File: metaflow/task.py (hunk 4 above)
Comment (darinyu) [reply metaflow/task.py:204]: I think its referring to the `not all_equal(lineage())` below.

File: metaflow/task.py (hunk 2 above)
Comment (darinyu) [reply metaflow/task.py:207]: we dont use index/value in this `lineage()`, so it should be the same giving them whatever value. (keeping it 0 to be consistent with index).

File: metaflow/task.py (hunk 3 above)
Comment (darinyu) [reply metaflow/task.py:480]: Chatted during meeting, This could be achieved but may be overkill. We also want people to be able to more easily access the foreach path, so putting them in metadata also makes sense.

I added a limit to how much data we store in metadata. If we encounter the arbitrary large string problem we can revisit this in the future.

File: metaflow/task.py [hunk 6]
```diff
@@ -24,7 +24,20 @@
...
+    foreach_frame_field_list,
+    defaults=(None,) * len(foreach_frame_field_list),
+)
+ForeachFrame.__reduce__ = lambda frame: (construct_frame, tuple(frame))
+# To be backwards compatible.
+def construct_frame(*args):
+    return ForeachFrame(*args[: len(ForeachFrame._fields)])
+
+
+# Maximum number of charactors of the foreach path that we store in the metadata.
```
Comment (romain-intel) [inline metaflow/task.py:39]: nit: characters.

File: metaflow/task.py (hunk 3 above)
Comment (romain-intel) [reply metaflow/task.py:480]: Yep -- and now that w have a limit of characters this should be fine.

File: metaflow/task.py [hunk 7]
```diff
@@ -463,6 +475,46 @@ def run_step(
...
+            foreach_stack = (
+                self.flow._foreach_stack
+                if hasattr(self.flow, "_foreach_stack") and self.flow._foreach_stack
+                else []
+            )
+
+            foreach_stack_formatted = []
+            current_foreach_path_length = 0
+            for frame in foreach_stack:
+                if not (frame.var and frame.var):
```
Comment (romain-intel) [inline metaflow/task.py:489]: frame.var and frame.value

**Review by romain-intel — APPROVED** [review]
LGTM with a small typo. Let's wait for someone from OB but we can merge afterwards.

File: metaflow/task.py (hunk 6 above)
Comment (darinyu) [reply metaflow/task.py]: done

File: metaflow/task.py (hunk 7 above)
Comment (darinyu) [reply metaflow/task.py]: done
